### ISSUE TITLE
Provide an official flatpak on flathub

### ISSUE THREAD

**Issue by Hofer-Julian** [gitbutlerapp/gitbutler#5776]
First, thanks a lot for your work. GitButler is fascinating, and I'd love to use it on a regular basis.

Since .deb packages are specific to Debian-based distributions and AppImage doesn't work on Wayland, I'd love to see if you'd take over the flatpak on [flathub](https://flathub.org/apps/com.gitbutler.gitbutler) and fix the remaining paper cuts.

The (seemingly) flatpak specific ones that I noticed are:
- git binary cannot be found (probably best to package it within the flatpak)
- it cannot access git repositories after restart (I would just give it access to the whole home folder)
- it attempts to acccess some D-Bus service (need to figure out which one and poke a whole for it)

I also hope that https://github.com/gitbutlerapp/gitbutler/pull/5609 helps here

**Comment by Byron** [comment]
Thanks a lot for the suggestion, and the helpful tips as well!

> * git binary cannot be found (probably best to package it within the flatpak)

Most Git clients I tried ended up doing this, even though a FlatPak release seems to be a special case of this as to GitButler, `git` would just natively appear in the `PATH` if it's bundled? This is more of a rhetorical question, as I know nothing about FlatPak really.

> * it attempts to acccess some D-Bus service (need to figure out which one and poke a whole for it)

This is probably `keyring` crate trying to read previously stored secrets from the system keychain. #5753 updates it to the latest version, which may also affect how this works on Linux.

CC @ndom91 who knows all about our FlatPak setup.

**Comment by ndom91** [comment]
Interesting timing haha. I've been working on a replacement flathub repo [here](https://github.com/ndom91/gitbutler-flathub) for the past few days. 

Building the app from scratch and bundling git seems to fix most issues. 

It seems to work building the debug build locally. However I can't get the release build to build locally yet, I think that's a more general rust / tauri issue with my build system though. The tauri build process just segfaults toward the end. 

The GHA is also not quite working yet, but that one seems more easily fixable. It can't find the appstream-compose binary. 

I've tried reaching out to the existing flathub gitbutler maintainer on our discord, they're a member, but haven't heard back.

**Comment by ndom91** [comment]
Also side note, the appimage doesn't work on wayland for you? I use it on Hyprland and another Ubuntu 24.04 gnome box regularly, what issues are you running into?

**Comment by Hofer-Julian** [comment]
@Byron 

> Most Git clients I tried ended up doing this, even though a FlatPak release seems to be a special case of this as to GitButler, git would just natively appear in the PATH if it's bundled? This is more of a rhetorical question, as I know nothing about FlatPak really.

Maybe you are more familiar with Docker? Flatpak behaves very similarly. Inside the container there's a filestructure resembling a classical unix file structure. If you put an executable into `/usr/bin` it will be in your PATH

**Comment by Hofer-Julian** [comment]
@ndom91 

> Interesting timing haha. I've been working on a replacement flathub repo [here](https://github.com/ndom91/gitbutler-flathub) for the past few days.

That's awesome. Let me know when you can need a second pair of eyes for something

> I've tried reaching out to the existing flathub gitbutler maintainer on our discord, they're a member, but haven't heard back.

Are you part of the GitButler team? If the current maintainer is unresponsive, don't hesitate to reach out to flathub maintainers. They will help to transfer ownership of the repository.

> Also side note, the appimage doesn't work on wayland for you? I use it on Hyprland and another Ubuntu 24.04 gnome box regularly, what issues are you running into?

The AppImage started on my Fedora machine. I don't know if it used Wayland or XWayland behind the scenes. I was actually under the impression that AppImage is completely incompatible with Wayland because of this [gist](https://gist.github.com/probonopd/9feb7c20257af5dd915e3a9f2d1f2277) by the AppImage creator. When trying it out, I've hit the mentioned clipboard problem, which is why I couldn't log into GitButler and therefore also not GitHub. And of course it doesn't install a desktop file and doesn't have an icon.

**Comment by ndom91** [comment]
Oh interesting, I'd never seen that gist. No so generally our appimage should "just work"😅 

I am on the gitbutler team, and yeah if it comes time to release it and I still haven't heard back then I'll reach out to the flathub team. 

Once I can get the release build building either locally or in CI I'd be happy about some more testing and feedback! I can ping you back in this issue with a download link if that's cool 🙏

**Comment by Hofer-Julian** [comment]
> Once I can get the release build building either locally or in CI I'd be happy about some more testing and feedback! I can ping you back in this issue with a download link if that's cool 🙏

Yes, please do that 🤝

**Comment by ndom91** [comment]
@Hofer-Julian alright so I've got it building the release build in CI finally. Unfortunately this still doesn't seem to communicate with the keyring correctly. I can login and everything works fine, however, as soon as I hard reload the application, or close / reopen it, the auth token is gone. 

Looking into the keyring, it doesn't look like anything is ever written into there.

Anyway, besides for that I haven't found any big issues yet: https://github.com/ndom91/gitbutler-flathub/actions/runs/12223161843

**Comment by Hofer-Julian** [comment]
Awesome!

Actually, I cannot reproduce your problems.
It stores both GitButler and GitHub credentials just fine for me.
If I restart the application, I am still logged in.

![image](https://github.com/user-attachments/assets/8e1c4ad6-e32b-44dd-9971-2eebbef28827)

**Comment by Hofer-Julian** [comment]
Copying link when logging into GitButler still doesn't work. Seems like I wrongly blamed AppImage for that. If I press on the whole field in order to trigger opening the browser instead, logging in works just fine.

**Comment by Byron** [comment]
> Looking into the keyring, it doesn't look like anything is ever written into there.

Maybe you could try again with #5753 applied? It updates the `keyring` crate and *maybe* that changes something to the better.

**Comment by ndom91** [comment]
@Hofer-Julian oh great news then haha. I'm testing this on a pretty vanilla Ubuntu 24.04 VM though so I can't be the only one having this issue :thinking: 

However, on my local nixos system I can't reproduce the keyring issue either. I can, however, reproduce the clipboard issue

![image](https://github.com/user-attachments/assets/e0274bdf-4706-49bd-94ba-4b56286a22d0)

@Byron I'll give your `keyring` crate PR a test tomorrow

**Comment by Hofer-Julian** [comment]
The copying is especially weird, since copying works just fine during the github authentication process.

**Comment by ndom91** [comment]
@Byron #5753 didn't seem to help in this case unfortunately :( 

That Ubuntu box still doesn't seem to write the info the password keyring. Also maybe notable, but its not working on both Ubuntu 24.04 VMs I have - one on my desktop at home and one on my laptop at work. They're both super vanilla gnome installations of 24.04.

**Comment by Hofer-Julian** [comment]
> That Ubuntu box still doesn't seem to write the info the password keyring. Also maybe notable, but its not working on both Ubuntu 24.04 VMs I have - one on my desktop at home and one on my laptop at work. They're both super vanilla gnome installations of 24.04.

Mmmh, but the AppImage or deb package adds the keyring entry correctly?

**Comment by ndom91** [comment]
> > That Ubuntu box still doesn't seem to write the info the password keyring. Also maybe notable, but its not working on both Ubuntu 24.04 VMs I have - one on my desktop at home and one on my laptop at work. They're both super vanilla gnome installations of 24.04.
> 
> Mmmh, but the AppImage or deb package adds the keyring entry correctly?

Correct and the same flatpak works well in my normal desktop nixos environment haha

Maybe something that Ubuntu is missing out of the box :thinking:

**Comment by Byron** [comment]
> @Byron #5753 didn't seem to help in this case unfortunately :(

Too bad, thanks for giving it a shot. But maybe there is still hope - there are [plenty of linux-features](https://docs.rs/keyring/latest/keyring/#credential-store-features) that one can enable to use different credential stores. If any of that makes sense to you, maybe that is something you could additionally try - it should be as easy as picking different cargo-features in the Cargo manifest that declares the `keyring` dependency.

**Comment by ndom91** [comment]
Yeah tbh I don't think its an issue on the side of our code / the `keyring` crate though at this point. Everywhere we've tried so far has worked (granted a small sample size :sweat_smile:) _except_ the two vanilla Ubuntu 24.04 VMs I have. 

So I'm thinking more that Ubuntu just doesn't come with somethign critical pre-installed. Pulling on that thread some more:

- It does have `gnome-keyring` installed as well as the gnome seahorse GUI for viewing your keyring.
- The AppImage / deb installations can interact with the keyring successfully on those same systems
- Maybe its some desktop portal / dbus communication issue? :thinking: 

Any other thoughts?

**Comment by Hofer-Julian** [comment]
One other thing one could try is this crate: https://crates.io/crates/oo7

On Linux it is actually preferred, since it stores the secret in a way that only GitButler can see it but not other apps.

**Comment by Byron** [comment]
Thanks for the suggestion, this is one of the coolest names for a crate for sure!

It seems Linux only and using it here would mean some effort as `keyring` is probably the best place to make it available in. And then we'd still not know if it makes a difference.

Trying different cargo features for `keyring` would be the first thing to try… but **trying the `cli` of `007` might be even easier**.

And speaking off, `keyring` also has [a CLI](https://github.com/hwchen/keyring-rs/blob/master/examples/cli.rs) - this should make trying different features trivial to find something that works consistently.

**Comment by ndom91** [comment]
Hey folks, a quick update - we're getting ready to release a first-party flatpak. I've been working on the build at https://github.com/ndom91/gitbutler-flathub. There you can find the latest build artifact (a `.flatpak` file you can install with `flatpak install ..`) under the github actions (i.e. https://github.com/ndom91/gitbutler-flathub/actions/runs/12789271694).

I've been in touch with the author of the community `com.gitbutler.gitbutler` [flathub package](https://flathub.org/apps/com.gitbutler.gitbutler) that's already out there and I will just commit our changes into that repository once we're ready and then going forward we'll maintain that together. 

So long story short, if you're already using the flathub package - it should work better and require no changes from you other than updating it.

We'd love any and all feedback you may have on it already, if you're up for trying out an experimental release, you can download it from the GHA linked above :pray:

**Comment by andre0xFF** [comment]
Hey. Just following up on this – what's the latest status? The Flatpak version (0.13.8) appears to be outdated

**Comment by Byron** [comment]
We aren't maintaining the flatpack distribution, so for now one would have to reach out to its maintainers for an update.

**Comment by Hofer-Julian** [comment]
> We aren't maintaining the flatpack distribution, so for now one would have to reach out to its maintainers for an update.

@ndom91 announced a few comments before that you are working on a first-party flatpak. I assume that's what @andre0xFF was referring to 🙂

**Comment by Byron** [comment]
Right, and thanks :D! This isn't actively worked on right now, but I imagine it increasing in priority once GitButler reaches version 1.0 .

**Comment by NN708** [comment]
Hi everyone, I'm also experiencing the same issues. it appears GitButler hasn't been updated on Flathub for over a year.

While @ndom91's PR (https://github.com/flathub/com.gitbutler.gitbutler/pull/104) is excellent, we haven't yet been able to resolve the issue with `pnpm`. I believe we should at least continue updating the package using the repackaging approach in the meantime. I've opened a PR here: https://github.com/flathub/com.gitbutler.gitbutler/pull/126, could someone please review it?

**Comment by Byron** [comment]
Thanks for taking the lead in this!

I did review and "approve" the linked PR, and maybe we can get this update out even before all problems that it may have are resolved. Also pinging @slarse as he probably uses Linux and is much closer to the tech-stack.

**Comment by slarse** [comment]
I don't think I've ever installed a flatpak, it's not really used much in my neck of the woods and wasn't around back when I was in the neck of the woods where it seems to be more popular. I'll definitely look into it now that I'm setting up a Debian machine to start working out compatibility kinks.

**Comment by NN708** [comment]
I've taken over the maintainership of the package on Flathub and have merged https://github.com/flathub/com.gitbutler.gitbutler/pull/126. If anyone from the GitButler team is interested in maintaining this package in the future, please fell free to reach out for access.

Some of the issues mentioned by @Hofer-Julian here remain unresolved, as they require changes to the default permissions. I've opened https://github.com/flathub-infra/flatpak-builder-lint/pull/772 to address this. In the meantime, you can manually grant the necessary permissions using a tool like [Flatseal](https://flathub.org/en/apps/com.github.tchx84.Flatseal).

**Comment by NN708** [comment]
With the merge of https://github.com/flathub/com.gitbutler.gitbutler/pull/127, all reported issues should now be resolved. Please give it a test and let me know if you find any other problems.

**Comment by Byron** [comment]
That's great to hear, thanks a lot! 

Let's close this issue for now, and open new ones should other problems be discovered.
Thanks everyone for their help, it was truly indispensable!

**Comment by Hofer-Julian** [comment]
I can confirm the problems I described above are fixed with the latest flatpak release.
It asked me for my name and email the first time, not sure if that is expected considering that is in my global gitconfig.

All in all, the flatpak is very much usable in its current state, thanks a lot @NN708!

**Comment by Byron** [comment]
> I can confirm the problems I described above are fixed with the latest flatpak release.
> It asked me for my name and email the first time, not sure if that is expected considering that is in my global gitconfig.

Thanks for sharing this! My guess is that flatpak is sandboxed enough for GitButler to not be able to read your global Git configuration, which probably contains your user settings. It would be interesting to learn where it put the user information that it queried from you, because apparently it can *write* that location.
I'd definitely love to understand what's going on there :).
